text/pvm.tex

```diff
@@ -536,43 +569,69 @@ \subsubsection{Instructions with Arguments of Three Registers}
   \thead{$\instructions_\imath$} & \thead{\textbf{Name}} & \thead{$\gas$} & \thead{\textbf{Mutations}} \\
   \midrule
   \endhead
-  8&\token{add}&0&$\reg'_D = (\reg_A + \reg_B) \bmod 2^{32}$\\ \mrule
-  20&\token{sub}&0&$\reg'_D = (\reg_A + 2^{32} - \reg_B) \bmod 2^{32}$\\ \mrule
-  23&\token{and}&0&$\forall i \in \N_{32} : \bits{\reg'_D}_i = \bits{\reg_A}_i \wedge \bits{\reg_B}_i$\\ \mrule
-  28&\token{xor}&0&$\forall i \in \N_{32} : \bits{\reg'_D}_i = \bits{\reg_A}_i \oplus \bits{\reg_B}_i$\\ \mrule
-  12&\token{or}&0&$\forall i \in \N_{32} : \bits{\reg'_D}_i = \bits{\reg_A}_i \vee \bits{\reg_B}_i$\\ \mrule
-  34&\token{mul}&0&$\reg'_D = (\reg_A \cdot \reg_B) \bmod 2^{32}$\\ \mrule
-  67&\token{mul\_upper\_s\_s}&0&$\reg'_D = \unsigned{\floor{(\signed{\reg_A} \cdot \signed{\reg_B}) \div 2^{32}}}$\\ \mrule
-  57&\token{mul\_upper\_u\_u}&0&$\reg'_D = \floor{(\reg_A \cdot \reg_B) \div 2^{32}}$\\ \mrule
-  81&\token{mul\_upper\_s\_u}&0&$\reg'_D = \unsigned{\floor{(\signed{\reg_A} \cdot \reg_B) \div 2^{32}}}$\\ \mrule
-  68&\token{div\_u}&0&$\reg'_D = \begin{cases}
-    2^{32} - 1 &\when \reg_B = 0\\
+  170&\token{add\_32}&0&$\reg'_D = \sext_4((\reg_A + \reg_B) \bmod 2^{32})$\\ \mrule
+  171&\token{sub\_32}&0&$\reg'_D = \sext_4((\reg_A + 2^{32} - \reg_B) \bmod 2^{32})$\\ \mrule
+  172&\token{mul\_32}&0&$\reg'_D = \sext_4((\reg_A \cdot \reg_B) \bmod 2^{32})$\\ \mrule
+  173&\token{div\_u\_32}&0&$\reg'_D = \begin{cases}
+    2^{64} - 1 &\when \reg_B = 0\\
+    \sext_4(\floor{\reg_A \div \reg_B}) &\otherwise
+  \end{cases}$\\ \mrule
+  174&\token{div\_s\_32}&0&$\reg'_D = \begin{cases}
+    2^{64} - 1 &\when \reg_B = 0\\
+    \reg_A &\when \signedn{4}{\reg_A} = -2^{31} \wedge \signedn{4}{\reg_B} = -1\\
+    \unsigned{\floor{\signedn{4}{\reg_A} \div \signedn{4}{\reg_B}}} &\otherwise
+  \end{cases}$\\ \mrule
+  175&\token{rem\_u\_32}&0&$\reg'_D = \begin{cases}
+    \reg_A &\when \reg_B = 0\\
+    \reg_A \bmod \reg_B &\otherwise
+  \end{cases}$\\ \mrule
+  176&\token{rem\_s\_32}&0&$\reg'_D = \begin{cases}
+    \reg_A &\when \reg_B = 0\\
+    0 &\when \signed{\reg_A} = -2^{31} \wedge \signed{\reg_B} = -1\\
+    \unsigned{\signed{\reg_A} \bmod \signed{\reg_B}} &\otherwise
+  \end{cases}$\\ \mrule
+  177&\token{shlo\_l\_32}&0&$\reg'_D = (\reg_A \cdot 2^{\reg_B \bmod 32}) \bmod 2^{32}$\\ \mrule
+  178&\token{shlo\_r\_32}&0&$\reg'_D = \floor{\reg_A \div 2^{\reg_B \bmod 32}}$\\ \mrule
+  179&\token{shar\_r\_32}&0&$\reg'_D = \unsigned{\floor{\signed{\reg_A} \div 2^{\reg_B \bmod 32}}}$\\ \mrule
```

Comment:
The sign-extension (into 64-bit) should be achieved through the `\unsigned{...}` when the operand is a signed value (in this case it is since we're using `\signed`)